src/test/java/org/jboss/remoting3/test/MaxChannelsTestCase.java

```diff
@@ -0,0 +1,207 @@
+import org.jboss.remoting3.spi.NetworkServerProvider;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+import org.wildfly.security.WildFlyElytronProvider;
+import org.wildfly.security.auth.client.AuthenticationConfiguration;
+import org.wildfly.security.auth.client.AuthenticationContext;
+import org.wildfly.security.auth.client.MatchRule;
+import org.wildfly.security.auth.realm.SimpleMapBackedSecurityRealm;
+import org.wildfly.security.auth.server.MechanismConfiguration;
+import org.wildfly.security.auth.server.SaslAuthenticationFactory;
+import org.wildfly.security.auth.server.SecurityDomain;
+import org.wildfly.security.password.PasswordFactory;
+import org.wildfly.security.password.spec.ClearPasswordSpec;
+import org.wildfly.security.permission.PermissionVerifier;
+import org.wildfly.security.sasl.SaslMechanismSelector;
+import org.wildfly.security.sasl.util.SaslMechanismInformation;
+import org.wildfly.security.sasl.util.ServiceLoaderSaslServerFactory;
+import org.xnio.IoFuture;
+import org.xnio.IoUtils;
+import org.xnio.OptionMap;
+import org.xnio.Options;
+import org.xnio.XnioWorker;
+
+import javax.net.ssl.SSLContext;
+import javax.security.sasl.SaslServerFactory;
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.security.PrivilegedAction;
+import java.security.Security;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests that the max inbound channels specified in default option map of EndPoint is honored.
+ *
+ * @author <a href="mailto:aoingl@gmail.com">Lin Gao</a>
+ */
+public class MaxChannelsTestCase {
+
+    protected final Endpoint clientEndpoint;
+    protected final Endpoint serverEndpoint;
+
+    private Closeable server;
+
+    private static String providerName;
+
+    @BeforeClass
+    public static void doBeforeClass() {
+        final WildFlyElytronProvider provider = new WildFlyElytronProvider();
+        Security.addProvider(provider);
+        providerName = provider.getName();
+    }
+
+    @AfterClass
+    public static void doAfterClass() {
+        Security.removeProvider(providerName);
+    }
+
+    public MaxChannelsTestCase() throws IOException {
+        clientEndpoint = Endpoint.builder().setDefaultConnectionsOptionMap(optionMap()).setEndpointName("connection-test-client").build();
+        serverEndpoint = Endpoint.builder().setDefaultConnectionsOptionMap(optionMap()).setEndpointName("connection-test-server").build();
```

Comment:
Due to the complicated way in which options are used in Remoting, with the current implementation (before any changes you made to Remoting) I think I can now make two statements about the way options are set in your otherwise good test case:
1. setting the defaultConnectionOptionsMap on the server endpoint has no effect on the configuration of incoming connection attempts; it only affects outgoing connection attempts, so configuring the map on the server endpoint has no effect on this test case and can be eliminated
2. the only way to pass options to affect incoming connection attempts is via the Map passed to NetworkServerProvider.createServer(...,optionMap,...) , so any options for handling incoming connections needs to be added there, in particular, you would want to pass in the options you set in optionMap(). 
3. MAX_INBOUND_CHANNELS and MAX_OUTBOUND_CHANNELS are connection establishment properties and so are correctly passed as in the comments 1. and 2.

Reply:
Thanks for the feedback, this is more clear to me. I have updated accordingly in https://github.com/tadamski/jboss-remoting/pull/2